Migrate project metadata to `pyproject.toml`

SpecLad opened the pull request:
PEP 621 seems to be the current best practice, so it would be good for pip to demonstrate it.

I didn't add a changelog entry, since the user-facing impact is negligible (see below). Tell me if I should add one anyway.

-----

Since setuptools defaults `include_package_data` to `True` if it sees a `project` table, explicitly set it to `False` to prevent unnecessary files from being added.

Due to differences between the semantics of `pyproject.toml` and `setup.py` settings, there are some minor changes in the resulting metadata:

    diff -Nur dist-old/whl/pip-23.1.dev0.dist-info/METADATA dist/whl/pip-23.1.dev0.dist-info/METADATA
    --- dist-old/whl/pip-23.1.dev0.dist-info/METADATA       2023-03-28 18:46:48.000000000 +0300
    +++ dist/whl/pip-23.1.dev0.dist-info/METADATA   2023-03-28 18:43:28.000000000 +0300
    @@ -2,10 +2,9 @@
     Name: pip
     Version: 23.1.dev0
     Summary: The PyPA recommended tool for installing Python packages.
    -Home-page: https://pip.pypa.io/
    -Author: The pip developers
    -Author-email: distutils-sig@python.org
    +Author-email: The pip developers <distutils-sig@python.org>
     License: MIT
    +Project-URL: Homepage, https://pip.pypa.io/
     Project-URL: Documentation, https://pip.pypa.io
     Project-URL: Source, https://github.com/pypa/pip
     Project-URL: Changelog, https://pip.pypa.io/en/stable/news/
    @@ -24,6 +23,7 @@
     Classifier: Programming Language :: Python :: Implementation :: CPython
     Classifier: Programming Language :: Python :: Implementation :: PyPy
     Requires-Python: >=3.7
    +Description-Content-Type: text/x-rst
     License-File: LICENSE.txt
     License-File: AUTHORS.txt

<!---
Thank you for your soon to be pull request. Before you submit this, please
double check to make sure that you've added a news file fragment. In pip we
generate our NEWS.rst from multiple news fragment files, and all pull requests
require either a news file fragment or a marker to indicate they don't require
one.

To read more about adding a news file fragment for your PR, please check out
our documentation at: https://pip.pypa.io/en/latest/development/contributing/#news-entries
-->

SpecLad commented:
Okay, I've no idea what's going on with those CI checks. I _did_ add a news fragment, and I didn't touch that file that black complains about. :/

webknjaz commented:
> I've no idea what's going on with those CI checks. I _did_ add a news fragment

The GitHub App based check is somewhat stricter — it expects a number in the name, pointing at an issue or a PR, not a UUID. @pradyunsg what do you want to do with this? Another feature toggle in the bot.

Regarding other PR aspects, I would argue that it's not a "trivial" change, changelog-wise. It definitely has a far-reaching effect so hiding this from the changelog is probably not an option, anyway.
Also, pip typically tries to maintain backwards compatibility longer than other projects. Thus, I have a hunch that the pip maintainers might not see such a change as acceptable at the moment.

SpecLad commented:
> Also, pip typically tries to maintain backwards compatibility longer than other projects. Thus, I have a hunch that the pip maintainers might not see such a change as acceptable at the moment.

What's the incompatibility here?

SpecLad commented:
> and I didn't touch that file that black complains about.

I figured this out. black infers the target Python version from the `requires-python` setting in `pyproject.toml`. In this case, it enables Python 3.7-specific formatting. This is actually a tangible benefit from switching to PEP 621!

I'll update the PR soon to fix the formatting.

pfmoore commented:
> In this case, it enables Python 3.7-specific formatting.

Huh? Why would *formatting* be version-specific? That makes no sense to me.

pradyunsg commented:
> Why would formatting be version-specific? That makes no sense to me.

Certain language syntax features are tied to Python versions, so formatting of long with statements for example changes based on whether you're on a version that allows parenthesis with with statements.

sbidoul commented:
If we do this, let's take the opportunity to move the remaining options in setup.py to `[tools.setuptools]` in `pyproject.toml` ? A quick look at the [relevant setuptools docs](https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html) seem to indicate it should be possible.

SpecLad commented:
> If we do this, let's take the opportunity to move the remaining options in setup.py to `[tools.setuptools]` in `pyproject.toml` ? A quick look at the [relevant setuptools docs](https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html) seems to indicate it should be possible.

According to that page, `tool.setuptools` is in beta stage, and I figured that a critical project like pip should not rely on beta functionality.

sbidoul commented:
Fair enough. In that case I wonder if it's not better to wait until setuptools declares the `tool.setuptools` section stable so we keep all our packaging config in one place?

SpecLad commented:
> In that case I wonder if it's not better to wait until setuptools declares the tool.setuptools section stable so we keep all our packaging config in one place?

The wait is now over, so I went ahead and moved the setuptools configuration into `tool.setuptools`.

I also caved and added a proper changelog entry.

sbidoul commented on pyproject.toml:
```diff
@@ -1,7 +1,72 @@
+  {name = "The pip developers", email = "distutils-sig@python.org"},
+]
+
+# NOTE: python_requires is duplicated in __pip-runner__.py.
```
```suggestion
# NOTE: requires-python is duplicated in __pip-runner__.py.
```

SpecLad replied:
Right. Fixed.

sbidoul commented on news/11909.feature.rst:
```diff
@@ -0,0 +1 @@
+Most project metadata is now defined statically in the ``pyproject.toml`` file.
```
Since this is not user-visible, adding this to the changelog might create confusion.

sbidoul replied:
Alternatively, this should be reformulated to make it more explicit that it is pip's own pyproject.toml.

SpecLad replied:
I tried to clarify it.

SpecLad replied:
FWIW, the reason I decided to make an entry after all is that this could _potentially_ be of interest to people using tools to examine sdist files. It's not an especially prominent change, but it is externally visible. 🤷‍♂️

sbidoul replied:
`11909.process.rst` is more appropriate than `feature` in this case.

sbidoul replied:
@SpecLad there is still the renaming of the newsfragment to do.

SpecLad replied:
Oh, sorry, I forgot to push this. Done now.

sbidoul commented on setup.py:
```diff
@@ -1,88 +1,13 @@
         "console_scripts": [
             "pip=pip._internal.cli.main:main",
             "pip{}=pip._internal.cli.main:main".format(sys.version_info[0]),
             "pip{}.{}=pip._internal.cli.main:main".format(*sys.version_info[:2]),
```
Does this `pip3.x` entry point still make sense?

In the pip wheel, it definitely does not, and actually varies from one release to the other, depending on which python version the maintainer doing the release was using.

I'd say we can make these entry points static and keep `pip` and `pip3` only.

SpecLad replied:
I agree in principle, but I don't think this PR is the right place to make this change. I specifically tried to avoid changing the metadata and focus on the format.

(Also, I received some pushback on this in https://github.com/pypa/pip/pull/11547#issuecomment-1296318295, and I don't want to add anything controversial to the PR.)

sbidoul replied:
I was not aware of that [hackery](https://github.com/pypa/pip/blob/ff05e4224bd7d1030e46d07e3a9c9419d39292fe/src/pip/_internal/operations/install/wheel.py#L275) to generate the pip console scripts. Interesting.

If I read that correctly, whatever `pip` console entry points we put in the wheels are ignored. So it's probably better to be correct and not generate the 3.x one which is going to be wrong most of the time.

> I agree in principle, but I don't think this PR is the right place to make this change. I specifically tried to avoid changing the metadata and focus on the format.

Yet this PR changes the behaviour of `setup.py`, right? if anyone was still using `setup.py install` to install pip, this PR will break it, possibly silently? So I'd say it's better to go all the way and remove `setup.py` instead of keeping a broken one.

SpecLad replied:
> Yet this PR changes the behaviour of `setup.py`, right? if anyone was still using `setup.py install` to install pip, this PR will break it, possibly silently?

I don't think so. I just tried it, seems to be working fine.

sbidoul replied:
Ok, that's right, setup.py reads the config from pyproject.toml, so that's good.

And keeping the 3.x entry point is valid for folks using, say, `build` + `installer` in the same environment. Although I still think we should aim at removing it that can be done in other PR.

SpecLad replied:
So, are we good now?

sbidoul reviewed (commented):
Looks good to me.

sbidoul commented:
> Regarding other PR aspects, I would argue that it's not a "trivial" change, changelog-wise. It definitely has a far-reaching effect so hiding this from the changelog is probably not an option, anyway.

@webknjaz what kind of issues do you anticipate here?

webknjaz commented:
> @webknjaz what kind of issues do you anticipate here?

@sbidoul mostly thinking of the downstream packagers that may be surprised too late in their cycles..

uranusjr commented:
Last time we did something similar there were significant backlashes from redistributors since many existing workflows rely on `setup.py`. I don’t know if things have changed, but be prepared to have this reverted if this turns out to be problematic again.

sbidoul commented:
At this stage setup.py is still there so it is probably fine?

pfmoore commented on tests/unit/test_network_auth.py:
```diff
@@ -406,7 +406,7 @@ def __call__(
         stdout: Optional[Any] = None,
         input: Optional[bytes] = None,
-        check: Optional[bool] = None
+        check: Optional[bool] = None,
```
Nit: Unrelated change. I don't *really* care, but if we did have to revert, it might be a nuisance to have this reverted as well.

SpecLad replied:
This change _is_ related, see https://github.com/pypa/pip/pull/11909#issuecomment-1493209124.

SpecLad commented:
I noticed that the 23.3 release cycle just closed. How about merging this now?
